tower-http/src/timeout/body.rs

```diff
@@ -0,0 +1,373 @@
+//!
+//! # Differences from `tower_http::timeout::service::Timeout`
+//!
+//! `tower_http::timeout::service::Timeout` applies a timeout on the full request.
+//! That timeout is not reset when bytes are handled, whether the request is active or not.
+//!
+//! This middleware will return a [`TimeoutError`].
+//!
+//! # Example
+//!
+//! ```
+//! use http::{Request, Response};
+//! use hyper::Body;
+//! use std::time::Duration;
+//! use tower::ServiceBuilder;
+//! use tower_http::timeout::body::RequestBodyTimeoutLayer;
+//!
+//! async fn handle(_: Request<Body>) -> Result<Response<Body>, std::convert::Infallible> {
+//!     // ...
+//!     # todo!()
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let svc = ServiceBuilder::new()
+//!     // Timeout bodies after 30 seconds of inactivity
+//!     .layer(RequestBodyTimeoutLayer::new(Duration::from_secs(30)))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+
+use futures_core::Future;
+use http::{Request, Response};
+use http_body::Body;
+use pin_project_lite::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{sleep, Sleep};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pin_project! {
+    /// Wrapper around a [`http_body::Body`] to time out if data is not ready within the specified duration.
+    pub struct TimeoutBody<B> {
+        timeout: Duration,
+        #[pin]
+        sleep: Option<Sleep>,
+        #[pin]
+        body: B,
+    }
+}
+
+impl<B> TimeoutBody<B> {
+    /// Creates a new [`TimeoutBody`].
+    pub fn new(timeout: Duration, body: B) -> Self {
+        TimeoutBody {
+            timeout,
+            sleep: None,
+            body,
+        }
+    }
+}
+
+impl<B> Body for TimeoutBody<B>
+where
+    B: Body,
+{
+    type Data = B::Data;
+    type Error = Box<dyn std::error::Error + Send + Sync>;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let mut this = self.project();
+
+        // Start the `Sleep` if not active.
+        let sleep_pinned = if let Some(some) = this.sleep.as_mut().as_pin_mut() {
+            some
+        } else {
+            this.sleep.set(Some(sleep(*this.timeout)));
+            this.sleep.as_mut().as_pin_mut().unwrap()
+        };
+
+        // Error if the timeout has expired.
+        match sleep_pinned.poll(cx) {
+            Poll::Pending => (),
+            Poll::Ready(()) => return Poll::Ready(Some(Err(Box::new(TimeoutError)))),
+        }
+
+        // Check for body data.
+        match this.body.poll_data(cx) {
+            Poll::Ready(data) => {
+                // Some data is ready. Reset the `Sleep`...
+                this.sleep.set(Some(sleep(*this.timeout)));
+
+                // ...then `poll` it to get awoken.
+                if let Poll::Ready(_) = this.sleep.as_pin_mut().unwrap().poll(cx) {
```

Comment:
Why do we set the sleep above then repoll it again? If we have data available we should return that right away and then on the next poll check if there is a timeout. That is if this timeout is a data read frame timeout aka we timeout if we've been pending for a data frame for too long which I think this is what you are trying to achieve.

Reply:
The rationale behind `poll`ing it as soon as it's set is that we want to register `cx` with the sleep reactor so that when the timeout is complete this task is woken. The fear is: if we don't `poll` here this task could sit indefinitely waiting for the data to `.wake()` up the task.

It seems that, if we don't care about the sleep reactor calling `.wake()` then we might as well use `Instant::now()`?

Reply:
Well I guess what is the timeout for? Is it for each data frame? So lets say I pull one data frame out then does the timer start then or when I want to pull the next one out. If I want to do something in between data frames this breaks that logic a bit, no?